### PR TITLE
Simplify watching for a settings changes

### PR DESCRIPTION
**Pull request by idegtiarenko**
There is a common pattern in code where we read a setting value during a service construction and also add a watcher for a future changes. This change:
* merges above 2 actions into a single call
* makes it unnecessary to pass a Settings into constructor via argument

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java [hunk 1]
```diff
@@ -130,19 +129,14 @@ public BalancedShardsAllocator(Settings settings) {
 
     @Inject
     public BalancedShardsAllocator(Settings settings, ClusterSettings clusterSettings, WriteLoadForecaster writeLoadForecaster) {
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:131]: There is a `settings` object in `clusterSettings` so it is no longer necessary to pass it into constructor.

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:131]: Makes sense. Why not remove this now-unused constructor argument?

**Review by DaveCTurner — COMMENTED** [review]
Looks good, I left some comments.

File: server/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java [hunk 2]
```diff
@@ -434,6 +434,17 @@ public synchronized <T> void addSettingsUpdateConsumer(Setting<T> setting, Consu
         addSettingsUpdateConsumer(setting, consumer, (s) -> {});
     }
 
+    /**
+     * This methods passes the setting value to a consumer during the initialization and on every setting change
+     * <p>
+     * Note: Only settings registered in {@link org.elasticsearch.cluster.ClusterModule} can be changed dynamically.
```
[the remark this answers is not in the document]
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java:440]: This was copied from `addSettingsUpdateConsumer` let me check that

**Review by DaveCTurner — APPROVED** [review]
LGTM. I really like that we no longer need to pass in a `Settings` and a `ClusterSettings` everywhere.

**Pull request merged**